### ISSUE TITLE
Index dtype changes when assigning column from index

### ISSUE THREAD

**Issue by jdanbrown** [dask/dask#2564]
- Creating a new ddf column from the ddf's existing index erroneously changes the dtype of the index
- Impact: causes downstream ddf operations to fail or behave in unintended ways, e.g. dd.concat

```py
>>> ddf = dd.from_pandas(pd.DataFrame(dict(i=['a','b'], a=[1,2])), npartitions=1)
>>> ddf = ddf.set_index('i)

>>> ddf.index

Dask Index Structure:
npartitions=1
a    object
b       ...
Name: i, dtype: object
Dask Name: sort_index, 4 tasks

>>> ddf.assign(b=ddf.index).index

Dask Index Structure:
npartitions=1
a    int64
b      ...
dtype: int64
Dask Name: assign, 6 tasks
```

## Versions
```sh
$ python --version
Python 3.6.0

$ cat requirements.txt | egrep 'dask/dask' | grep -v '^#'
git+https://github.com/dask/dask.git@00ed56e
```

**Comment by jdanbrown** [comment]
cc @bkyryliuk @brianmartin @xinluh

**Comment by TomAugspurger** [comment]
Potentially a pandas issue:

```python

In [24]: pd.DataFrame().assign(i=[]).index.dtype
Out[24]: dtype('int64')

In [25]: pd.DataFrame().index.dtype
Out[25]: dtype('O')

In [26]: pd.DataFrame().assign(i=[]).index.dtype
Out[26]: dtype('int64')
```

which boils down to

```python
In [33]: df = pd.DataFrame(columns=['A'])

In [34]: df.index.dtype
Out[34]: dtype('O')

In [35]: df['A'] = []

In [36]: df.index.dtype
Out[36]: dtype('int64')

In [37]: df.index
Out[37]: RangeIndex(start=0, stop=0, step=1)
```

**Comment by jimfulton** [comment]
Here's another example:

```
>>> df = pd.DataFrame([dict(x='c', y='a'), dict(x='d', y='b')])
>>> ddf = dd.from_pandas(df, 2)
>>> ddf.dtypes
x    object
y    object
dtype: object
>>> ddf.rename(columns=dict(x='index')).groupby('index')['y'].apply('-'.join, meta=str).reset_index().dtypes
index     int64
0        object
dtype: object
```

This prevents applying string operations to the variable. 

Now, if I add compute:
```
>>> ddf.rename(columns=dict(x='index')).groupby('index')['y'].apply('-'.join, meta=str).reset_index().compute().dtypes
index    object
0        object
dtype: object
```

We get the correct value.

BTW, I don't see this problem with pandas.
```
>>> df.rename(columns=dict(x='index')).groupby('index')['y'].apply('-'.join).reset_index().dtypes
index    object
y        object
dtype: object
```

So while something about pandas meta data may confuse dask, it doesn't prevent the corresponding operations from working in pandas.

**Comment by jcrist** [comment]
This is fixed by #4695:

```python
In [1]: %load test.py

In [2]: # %load test.py
   ...: import pandas as pd
   ...: import dask.dataframe as dd
   ...:
   ...: df = pd.DataFrame(dict(i=['a','b'], a=[1,2]))
   ...: ddf = dd.from_pandas(df, npartitions=1).set_index('i')
   ...: ddf2 = ddf.assign(b=ddf.index)
   ...: assert ddf2.index.dtype == object
   ...:

In [3]: %load test2.py

In [4]: # %load test2.py
   ...: import pandas as pd
   ...: import dask.dataframe as dd
   ...:
   ...: df = pd.DataFrame([dict(x='c', y='a'), dict(x='d', y='b')])
   ...: ddf = dd.from_pandas(df, 2)
   ...: ddf2 = ddf.rename(columns=dict(x='index'))
   ...: ddf3 = ddf2.groupby('index')['y'].apply('-'.join, meta=str)
   ...: ddf4 = ddf3.reset_index()
   ...: assert ddf4['index'].dtype == 'int64'
   ...:
```

**Comment by jdanbrown** [comment]
@jcrist Thanks!
